design/rest-api-guideline.md

```diff
@@ -0,0 +1,389 @@
+
+The `version` refers to the stability and the state of an API. The version does not necessarily refers to a Keycloak release version number so that it can be mapped to different Keycloak release versions.
+
+Here is an example of a resource path related with the group of functionalities provided by the Account Management REST API:
+
+```
+/apis/accounts/v1/{user_id}
+```
+
+From the example above we can infer that the path refers to a specific `v1` version with the possibility to change the version and: 
+
+* completely change the API
+* do breaking changes to parts of the API
+* remove functionalities or specific parts of the API
+* provide non-stable or technology preview features
+* the different APIs can evolve separately
+
+#### Extensions
+
+To better acommodate and separate the core API from extensions built on top of the API, developers should consider the following path format for extensions:
+
+```
+/apis/{API_GROUP}/extensions/{version}
+```
+
+This separation aims to avoid confusion when consuming stable and official Keycloak APIs so that clients can differentiate if they are using some API provided by a third-party with no support from Keycloak team.
+
+#### Technology Preview
+
+To better acommodate and separate the core API from technology preview features, developers should use a beta version as the API version:
+
+```
+/apis/{API_GROUP}/v1beta1
+```
+
+Technology preview APIs are related with features that have great potential to become fully supported. They provide the latest work in some area of interest and are offered with the intent to gather feedback from the community about new functionalities.
+
+#### Store Resources
+
+A resource store provides operations for a given resource. For example:
+
+```
+/apis/accounts/v1/1234/applications
+```
+
+The example above is representing a resource store path from where we can manage the `applications` for a user account with ID `1234`. Based on this path, clients can perform different operations based on the HTTP verbs or any controller resource defined to the store. Usually this type of resource is associated with CRUD operations for a specific resource.
+
+When a resource fits under this category, the name should be a plural noun.
+
+#### Controller Resources
+
+Under some circumstances, the HTTP verbs are not enough to represent an operation that can be performed in a resource. In this case,
+the resource path may indicate a specific action to be taken when accessed.
+
+```
+/apis/accounts/v1/1234/disable
+```
+
+The example above shows the usage of `disable` to operate accounts in order to disable them.
+
+When a resource fits under this category, the name should be a verb, using camel-case if necessary.
+
+### Resource Representation
+
+The standard format for resource representation is JSON. There is no rule for schemas or the semantics used by a resource representation, as long as they follow the JSON syntax properly.
```

Comment:
Good point. To avoid confusion whether or not HATEOAS should be considered, I've included a `Level of Service Maturity` section that states we should be doing Level 2. Pretty much what we do today, no HATEOAS.

Reply:
Thanks for adding that section. I believe the HATEOAS question is clear now.

Reply:
Thank you for all the valuable feedback.